RE: Add Developer's Comments to add-on detail/description

xlisachan opened the pull request:
Fixes #2764 

This patch adds the developer's comments to each add-on if it's available.

@willdurand , please review when you have a moment and let me know if you have any feedback. Thank you!

Before - Awesome Screenshot Plus
<img width="1000" alt="Screen Shot 2020-03-20 at 7 22 33 PM" src="https://user-images.githubusercontent.com/28129806/77213497-643b1180-6ae1-11ea-940b-e844ac86e2c1.png">

After - Awesome Screenshot Plus
<img width="1000" alt="Screen Shot 2020-03-20 at 7 02 40 PM" src="https://user-images.githubusercontent.com/28129806/77213481-584f4f80-6ae1-11ea-98bb-5363925680ee.png">

willdurand commented on src/amo/pages/Addon/index.js:
```diff
@@ -299,6 +299,34 @@ export class AddonBase extends React.Component {
+    }
+
+    if (!comments || comments === null) {
+      showComments = false;
```
I think we could omit the `showComments` variable if we use a early return statement:

```js
if (!comments) {
  return null;
}
```

willdurand commented on src/amo/pages/Addon/index.js:
```diff
@@ -299,6 +299,34 @@ export class AddonBase extends React.Component {
+    if (!comments || comments === null) {
+      showComments = false;
+    }
+    commentsProps.dangerouslySetInnerHTML = sanitizeUserHTML(comments);
```
I think you can inline this directly on the component (`div`)

willdurand commented on src/amo/pages/Addon/index.js:
```diff
@@ -299,6 +299,34 @@ export class AddonBase extends React.Component {
+    return showComments ? (
+      <ShowMoreCard
+        className={showMoreCardName}
+        header={`Developer's comments`}
```
We will need to translate this string so we should call `i18n.gettext()`:

```suggestion
        header={i18n.gettext(`Developer's comments`)}
```

xlisachan commented:
@willdurand Thanks for the feedback! I made the requested updates and added a test to check that developer's comments are hidden if null. Please let me know if additional tests are required or whether that suffices when you have a moment.

willdurand commented on src/amo/pages/Addon/index.js:
```diff
@@ -299,6 +299,34 @@ export class AddonBase extends React.Component {
+        header={i18n.gettext(`Developer's comments`)}
+        id={showMoreCardName}
+      >
+        <div dangerouslySetInnerHTML={devComments} />
```
So maybe we could simplify the code like this:

```suggestion
    if (!addon || !addon.developer_comments) {
      return null;
    }
    
    const devComments = sanitizeUserHTML(addon.developer_comments);
    const showMoreCardName = 'AddonDevelopersComments';

    /* eslint-disable react/no-danger */
    return (
      <ShowMoreCard
        className={showMoreCardName}
        header={i18n.gettext(`Developer's comments`)}
        id={showMoreCardName}
      >
        <div dangerouslySetInnerHTML={devComments} />
```

willdurand commented on src/amo/pages/Addon/index.js:
```diff
@@ -299,6 +299,34 @@ export class AddonBase extends React.Component {
+    }
+    const devComments = sanitizeUserHTML(comments);
+
+    const showMoreCardName = 'AddonDevelopersComments';
```
```suggestion
    const showMoreCardName = 'Addon-developer-comments';
```

willdurand commented on tests/unit/amo/pages/TestAddon.js:
```diff
@@ -883,6 +883,15 @@ describe(__filename, () => {
+      ...fakeAddon,
+      developer_comments: null,
+    });
+    const root = renderAsDOMNode({ addon });
```
So it would be better to use `renderComponent()` in new test cases.

willdurand commented on tests/unit/amo/pages/TestAddon.js:
```diff
@@ -883,6 +883,15 @@ describe(__filename, () => {
     ).toContain('<a>placeholder</a>');
   });
 
+  it('hides developers comments if null', () => {
```
This is a good test case. Then we'll need at least one test case for when the add-on has a developer comment.

willdurand commented:
It looks very great so far, thanks! I noticed a UI issue, though but it is not a new one, see this screenshot:

![Screen Shot 2020-03-23 at 12 34 53](https://user-images.githubusercontent.com/217628/77312710-d0489000-6d02-11ea-8c4d-b441288049f8.png)

There is a problem with the card.

willdurand commented:
> There is a problem with the card.

For this problem, we use some style in `ShowMoreCard/styles.scss` (line 39). I suggest we add the class name for this new component here as well. Something like:

```
.AddonDescription,
.AddonDevelopersComments {
   ...
```

xlisachan commented:
Thanks for the additional feedback, @willdurand ! Per your request, I made the following updates:
- cleaned up index.js
- updated class from `.AddonDevelopersComments` to `.Addon-developer-comments`
- utilized `renderComponent()` in the test case
- added a test case for when developer comments exists
- revised stylesheet

<img width="724" alt="Screen Shot 2020-03-23 at 10 53 53 AM" src="https://user-images.githubusercontent.com/28129806/77331828-a178ec80-6cf7-11ea-905f-41ab7870b6ec.png">

willdurand commented on src/amo/pages/Addon/index.js:
```diff
@@ -299,6 +299,27 @@ export class AddonBase extends React.Component {
+  renderDevCommentsCard = () => {
+    const { addon, i18n } = this.props;
+
+    if (!addon || !addon.developer_comments) return null;
```
```suggestion
    if (!addon || !addon.developer_comments) {
      return null;
    }
```

willdurand commented on src/amo/pages/Addon/index.js:
```diff
@@ -299,6 +299,27 @@ export class AddonBase extends React.Component {
+    return (
+      <ShowMoreCard
+        className={showMoreCardName}
+        header={i18n.gettext(`Developer's comments`)}
```
@jvillalobos is this header title good? Or should we use `Author`? Or even `Developer`/`Artist`/`Author` depending on the context. What we should have when there are more than author?

jvillalobos replied:
I think `Developer comments` without the possessive is better because it's more generic (in English, at least).

willdurand commented on tests/unit/amo/pages/TestAddon.js:
```diff
@@ -872,6 +872,27 @@ describe(__filename, () => {
     );
   });
 
+  it('hides developers comments if null', () => {
```
```suggestion
  it(`hides the developer's comments if null`, () => {
```

willdurand commented on tests/unit/amo/pages/TestAddon.js:
```diff
@@ -872,6 +872,27 @@ describe(__filename, () => {
+    expect(root.find('.Addon-developer-comments')).toHaveLength(0);
+  });
+
+  it('displays developers comments', () => {
```
```suggestion
  it(`displays the developer's comments`, () => {
```

willdurand commented on tests/unit/amo/pages/TestAddon.js:
```diff
@@ -872,6 +872,27 @@ describe(__filename, () => {
+      }),
+    });
+    const devComments = root.find('.Addon-developer-comments div');
+    expect(devComments.html()).toContain(developerComments);
```
I would suggest:

```suggestion
    expect(root.find('.Addon-developer-comments').childAt(0).html()).toContain(
      developerComments,
    );
```

willdurand commented on tests/unit/amo/pages/TestAddon.js:
```diff
@@ -872,6 +872,27 @@ describe(__filename, () => {
+    const devComments = root.find('.Addon-developer-comments div');
+    expect(devComments.html()).toContain(developerComments);
+  });
+
```
It would be nice to have a test case similar to the one named `'allows some HTML tags in the description'`

xlisachan commented:
Per your request, I cleaned up the code and test case, and added a new test case to check that HTML tags are allowed.  I have also removed the possessive from developer in the header and test cases. Please review and let me know you thoughts. Thank you!

willdurand commented:
Impressive work @xlisachan, thanks!
